### ISSUE TITLE
KickAsync Leads to HTTPException

### ISSUE THREAD

**Issue by Tzadiko** [RogueException/Discord.Net#1446]
Hello,

I send this function `KickUser` the user I want kicked, the user doing the kicking, and the reason, if any.
```
[RequireBotPermission(GuildPermission.KickMembers)]
private async Task KickUser(IGuildUser user, SocketGuildUser personKicking, string reason = null)
{
    // user is the person being kicked.
    var textChannel = this._client.GetChannel(this._channelId) as SocketTextChannel;
    if (user == null) await textChannel.SendMessageAsync($"User doesn't exist. " +
        $"Cannot kick this user.");
    var role = user.Guild.Roles.FirstOrDefault(x => x.Name == "Admin");

    await textChannel.SendMessageAsync($"Attempting to kick {user.Mention}. " +
        $"Reason: {reason ?? "Admin has his reasons."}");

    if (!(user as SocketGuildUser).Roles.Contains(role))
    {
        if (personKicking.GuildPermissions.KickMembers)
        {
            await user.KickAsync(reason); // ERROR
        }
        else
        {
            await textChannel.SendMessageAsync($"You don't have permissions" +
                $" to kick {user.Mention}");
        }
    }
    else
    {
        await textChannel.SendMessageAsync($"{user.Mention} is an Admin." +
            $" They cannot be kicked.");
    }
}
```
Everytime it hits the "KickAsync" line, it throws an the following exception: `Exception thrown: 'Discord.Net.HttpException' in System.Private.CoreLib.dll`

Why is this happening? How can I resolve it?

**Comment by Playwo** [comment]
The only reason that would be obvious is that you are missing the permission to kick the user. 
The KickMembers Permission allows the bot to kick people who are lower in the role hierarchy than the bot. Moreover you can't kick the guild owner.

**Comment by k-boyle** [comment]
Pro tip, include the stack trace with the exception name

**Comment by Tzadiko** [comment]
@k-boyle Sorry, was late last night and completely forgot to post it.

`
HTTP Code: Forbidden. Stack Trace:    at Discord.Net.Queue.RequestBucket.SendAsync(RestRequest request)
   at Discord.Net.Queue.RequestQueue.SendAsync(RestRequest request)
   at Discord.API.DiscordRestApiClient.SendInternalAsync(String method, String endpoint, RestRequest request)
   at Discord.API.DiscordRestApiClient.SendAsync(String method, String endpoint, String bucketId, ClientBucketType clientBucket, RequestOptions options)
   at Discord.API.DiscordRestApiClient.RemoveGuildMemberAsync(UInt64 guildId, UInt64 userId, String reason, RequestOptions options)
   at Discord.Rest.UserHelper.KickAsync(IGuildUser user, BaseDiscordClient client, String reason, RequestOptions options)
   at Bot.Welcome.WelcomeBot.KickUser(IGuildUser user, SocketGuildUser personKicking, String reason) in C:\Users\Tomer\Documents\Coding\C#\Applications\DiscordBot\DiscordBot\Bot.Welcome\WelcomeBot.cs:line 215
`

**Comment by FiniteReality** [comment]
Forbidden means you don't have permission to do the thing you want to do. Probably because the user has a higher role than your bot.

EDIT: see https://discordapp.com/developers/docs/topics/opcodes-and-status-codes#http

**Comment by Tzadiko** [comment]
> Forbidden means you don't have permission to do the thing you want to do. Probably because the user has a higher role than your bot.
> 
> EDIT: see https://discordapp.com/developers/docs/topics/opcodes-and-status-codes#http

That's so strange, I added my bot with:

https://discordapp.com/oauth2/authorize?client_id=674106189239287818&scope=bot&permissions=8

The person she is attempting to kick is assigned the role of Newbie, which doesn't have the Admin role (let alone any Admin-related privileges, like kicking users).

Hmmmm...

![](https://i.gyazo.com/352936928b1762735ee38d5102f8b3f2.png)

**Comment by AntiTcb** [comment]
The role hierarchy doesn't care about permissions, it cares about the order a role is compared to another role on the list.

If your bot's role is underneath Newbie on the list, then the bot can't do anything to anyone with the Newbie role, even though it was granted administrative permission.

**Comment by FiniteReality** [comment]
Make sure the position of the role is higher - kicking/banning works on role position in the role list rather than permissions.

**Comment by Tzadiko** [comment]
Aaahhh that's how roles work! Haha, okay, I'm new to Discord and I skipped over the caption on the _Role Linear Hierachy_ picture here: https://support.discordapp.com/hc/en-us/articles/214836687-Role-Management-101

> The blurple arrow is your highest role; roles now follow a linear hierarchy. In simpler terms, the "metaphorical totem pole" is now a bit more literal in practice.

Thank you all for your prompt responses and help. You have a very intuitive and easy to use API.

![image](https://user-images.githubusercontent.com/44016383/74048941-2951a400-49a1-11ea-9357-c2168c3f5a82.png)
